pecos/core/ann/hnsw.hpp

```diff
@@ -238,6 +243,143 @@ namespace ann {
+
+        ProductQuantizer4Bits quantizer;
+        index_type num_node;
+        index_type code_dimension;
+        size_t code_offset;
+        size_t node_mem_size;
+        index_type max_degree;
+        std::vector<uint64_t> mem_start_of_node;
+        std::vector<char> buffer;
+
+        void save(FILE *fp) const {
+            pecos::file_util::fput_multiple<index_type>(&num_node, 1, fp);
+            pecos::file_util::fput_multiple<index_type>(&code_dimension, 1, fp);
+            pecos::file_util::fput_multiple<size_t>(&code_offset, 1, fp);
+            pecos::file_util::fput_multiple<size_t>(&node_mem_size, 1, fp);
+            pecos::file_util::fput_multiple<index_type>(&max_degree, 1, fp);
+            size_t sz = mem_start_of_node.size();
+            pecos::file_util::fput_multiple<size_t>(&sz, 1, fp);
+            if (sz) {
+                pecos::file_util::fput_multiple<uint64_t>(&mem_start_of_node[0], sz, fp);
+            }
+            sz = buffer.size();
+            pecos::file_util::fput_multiple<size_t>(&sz, 1, fp);
+            if (sz) {
+                pecos::file_util::fput_multiple<char>(&buffer[0], sz, fp);
+            }
+            quantizer.save(fp);
+            fclose(fp);
+        }
+
+        void load(FILE *fp) {
+            pecos::file_util::fget_multiple<index_type>(&num_node, 1, fp);
+            pecos::file_util::fget_multiple<index_type>(&code_dimension, 1, fp);
+            pecos::file_util::fget_multiple<size_t>(&code_offset, 1, fp);
+            pecos::file_util::fget_multiple<size_t>(&node_mem_size, 1, fp);
+            pecos::file_util::fget_multiple<index_type>(&max_degree, 1, fp);
+            size_t sz = 0;
+            pecos::file_util::fget_multiple<size_t>(&sz, 1, fp);
+            mem_start_of_node.resize(sz);
+            if (sz) {
+                pecos::file_util::fget_multiple<uint64_t>(&mem_start_of_node[0], sz, fp);
+            }
+            pecos::file_util::fget_multiple<size_t>(&sz, 1, fp);
+            buffer.resize(sz);
+            if (sz) {
+                pecos::file_util::fget_multiple<char>(&buffer[0], sz, fp);
+            }
+
+            quantizer.load(fp);
+
+            fclose(fp);
+        }
+
+
+        void build_quantizer(const pecos::drm_t& X_trn, index_type subspace_dimension, index_type sub_sample_points) {
+            index_type code_dimension = X_trn.cols;
+            if (subspace_dimension == 0) {
+                if (code_dimension >= 400) {
+                    code_dimension =  code_dimension % 2 == 0 ? code_dimension / 2 : code_dimension / 2 + 1;
+                }
+            } else {
+                code_dimension = code_dimension / subspace_dimension;
+            }
+            // currently, we don't support padding 0 on X_trn, so the cols of X_trn must be divisible by subspace_dimension.
+            // otherwise, we will throw error in quantizer.train().
+            quantizer.train(X_trn, code_dimension, sub_sample_points);
```

Comment:
in this case, `code_dimension` is actually `num_local_codebook` ?!

Reply:
consider naming variables for consistency in `quantizer.hpp`

Reply:
code_dimension is from the perspective of HNSWPQ4Bits and num_local_codebook is from the perspective of Quantizer.

Reply:
After reading again the code, I think renaming code_dimension into num_local_codebook_pad is actually not a good idea. From Graph construction perspective, it's the number of dimensions a code will use, and there is no concept of "local codebook" from Graph's view. In addition, when calculating the memory space for neighboring code, code_dimension is a much more meaningful name than num_local_codebook_pad. So I think the name could be kept same.

Reply:
I see. In such case, I would suggest you write some comments that describe the definitions of those variables under those structures (e.g., GraphL0PQ4bit, HNSWProdQuad4Bit, quantizer)

Reply:
Revise in the next commit.